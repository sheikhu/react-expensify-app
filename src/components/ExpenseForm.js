import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

import { addExpense } from '../actions/expenses';

export default class ExpenseForm extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: undefined,
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Description and Amount are required !' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  }
  onFocusChange = ({ focused }) => this.setState({ calendarFocused: focused })

  render() {
    return (
      <div>
        <h1>Expense Form</h1>

        {this.state.error && <p className="error">{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <div>
            <input
              type="text"
              placeholder="Description"
              autoFocus
              value={this.state.description}
              onChange={this.onDescriptionChange}
            />
          </div>

          <div>
            <input
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
            />
          </div>
          <div>
            <SingleDatePicker
              date={this.state.createdAt} // momentPropTypes.momentObj or null
              onDateChange={this.onDateChange} // PropTypes.func.isRequired
              focused={this.state.calendarFocused} // PropTypes.bool
              onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
          <div>
            <textarea
              placeholder="Add a note for your expense (optional)"
              value={this.state.note}
              onChange={this.onNoteChange}
            >
            </textarea>
          </div>

          <div>
            <button>Save</button>
          </div>
        </form>
      </div>
    );
  }
};

