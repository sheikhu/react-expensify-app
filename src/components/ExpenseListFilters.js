import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from 'react-dates';


export class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
          <div className="input-group">
              <div className="input-group-item">
                  <input
                      className="text-input"
                      type="text"
                      value={this.props.filters.text}
                      onChange={this.onTextChange}
                      placeholder="Search expenses"
                  />
              </div>
              <div className="input-group-item">
                  <select className="select"
                      value={this.props.filters.sortBy}
                      onChange={this.onSortChange}>
                      <option value=""> </option>
                      <option value="date">Date</option>
                      <option value="amount">Amount</option>
                  </select>
              </div>
              <div className="input-group-item">
                  <DateRangePicker
                      startDateId="startDate"
                      endDateId="endDate"
                      startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                      endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                      onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                      focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                      onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                      isOutsideRange={() => (false)}
                      numberOfMonths={1}
                      showClearDates={true}
                  />
              </div>
          </div>
      </div>
    );
  }
}


const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
