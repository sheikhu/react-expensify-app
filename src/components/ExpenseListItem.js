import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => (
  <div>
    <h3>
      <Link to={`/edit/${props.expense.id}`}>{props.expense.description}</Link>
    </h3>
    <p>
      {numeral(props.expense.amount / 100).format('$0,0.00')}
      |
      {moment(props.expense.createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;