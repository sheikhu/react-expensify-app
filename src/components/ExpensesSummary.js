import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpensesSummary extends React.Component {

  render() {
    const { expenseCount, expensesTotal } = this.props;
    const expenseWord = expenseCount == 1 ? 'expense' : 'expenses';
    return (
      <div>
        Viewing {expenseCount} {expenseWord}
        totaling {numeral(expensesTotal / 100).format('$0,0.00')}
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)

  };
};
export default connect(mapStateToProps)(ExpensesSummary);