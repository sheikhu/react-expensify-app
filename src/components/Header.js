import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLogout, startLogin } from '../actions/auth';

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
        <div className="header__content">
        <Link className="header__title" exact to="/dashboard" activeClassName="active">
            <h1>Expensify</h1>
        </Link>
        <button className="button button--link" onClick={startLogout}>Logout</button>
        </div>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);