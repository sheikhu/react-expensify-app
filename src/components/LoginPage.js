import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export class LoginPage extends React.Component {

  login = (e) => {
    e.preventDefault();

    this.props.startLogin();
    this.props.history.push('/dashboard');
  }

  render() {
    return (
      <div className="box-layout">
        <div className="box-layout__box">
          <h1 className="box-layout__title">Expensify</h1>
          <p>It's time to get your expenses under control.</p>
          <button className="button" type="button" onClick={this.props.startLogin}>Login with Google</button>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});
export default connect(undefined, mapDispatchToProps)(LoginPage);