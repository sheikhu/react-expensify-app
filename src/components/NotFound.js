import React from 'react';
import { NavLink } from 'react-router-dom';

const NotFound = () => (
  <div>
    Page Not found - <NavLink to="/" exact>Go Home</NavLink>
  </div>
);

export default NotFound;