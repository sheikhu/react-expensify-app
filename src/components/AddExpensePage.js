import React, { PropTypes } from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ReduxSweetAlert, { swal } from 'react-redux-sweetalert'; // eslint-disable-line

export class AddExpensePage extends React.Component {


  constructor(props) {
    super(props);
  }

  addExpense = (expense) => {
    this.props.startAddExpense(expense).then(() => {
      this.props.history.push('/dashboard');
    });

  };

  render() {
    return (
      <div>
        <ReduxSweetAlert show={true} />
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={this.addExpense}
          />
        </div>
      </div>
    );
  }
}


const mapStateToProps = (state) => ({});

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense)),
    swal
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
