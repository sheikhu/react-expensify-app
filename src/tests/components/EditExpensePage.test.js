import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';


let startEditExpense, startRemoveExpense, history, wrapper, expense;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage expense={expense} startEditExpense={startEditExpense} startRemoveExpense={startRemoveExpense} history={history} />);
    expense = expenses[0];
});

test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});


test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expense);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expense.id, expense);
});


test('should handle delete', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith(expense.id);
});