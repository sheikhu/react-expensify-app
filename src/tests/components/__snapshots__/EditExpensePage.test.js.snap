// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
    <button
      class="button button--secondary"
      onClick={[Function]}
      type="button"
    >
      Remove
    </button>
  </div>
</div>
`;
